### DISCUSSION TITLE
Performance: Tauri IPC vs React Native JSI

### DISCUSSION THREAD

**Discussion by velocitysystems** [tauri-apps/tauri#11915] [General]
Hello! 👋 I’m currently evaluating Tauri to replace React Native for a cross platform iOS, Android and Windows app. 

Is the performance of Tauri’s IPC a bottleneck for high-performance applications? E.g. Passing [large data payloads from SQL](https://github.com/tauri-apps/plugins-workspace/issues/2049) over the IPC is reported to be slow.

The Reactive Native Bridge (legacy) uses a similiar design where data crossing the bridge must be serialized/deserialized, however the [new JSI](https://reactnative.dev/architecture/landing-page#fast-javascriptnative-interfacing) facilitates _direct_ communication between JavaScript and native code bypassing any need for data serialization.

Putting aside architectural differences and security concerns, is there any way of achieving **similar performance** with Tauri's IPC?

**Comment by FabianLars** [comment]
> Putting aside architectural differences and security concerns, is there any way of achieving similar performance with Tauri's IPC?

Tauri has a serialization-free ipc in v2 too but it still uses the fetch api (without actually hitting the network) so there's a bit of a delay. A user [on discord](https://discord.com/channels/616186924390023171/1306724593549181070) did some non-scientific benchmark for the binary ipc and got ~5ms on macOS but ~200ms on Windows (that 100% used to be better so idk what's going on there), for 10mb of data.


With the currently used system webviews i believe that we can't improve it much further. Only the windows webview supports shared memory but webview2 users reported that it's weirdly slow (also ignoring that it's not exactly trivial to turn that into a usable api). If that's fixed we'd still be left with no macos/linux/android/ios support.

There are also sync webview apis that sound similar-ish to JSI (exposing a rust interface to js) but it's not really fitted for large data transfers either.

**Reply by Horbin-Magician** [reply]
Is there any way to improve the IPC in Windows? (~200ms for 10mb is too slow for me)

**Comment by velocitysystems** [comment]
Thanks @FabianLars. With that in mind what is the recommended path for large data transfers over the IPC? i.e. Transferring image data, large byte arrays etc?

**Reply by FabianLars** [reply]
https://v2.tauri.app/develop/calling-rust/#returning-array-buffers (rust->js) and https://v2.tauri.app/develop/calling-rust/#accessing-raw-request (js->rust) and maybeee https://v2.tauri.app/develop/calling-rust/#channels i think they shouldn't be much slower.
